[Custom Fields] Rename OrderMetaData entity to MetaData

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #13619
Please note that while this PR changes quite a bit of files, most of them are just string replacements.

<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR is meant to rename and migrate the existing `OrderMetaData` entity to `MetaData`, while also moving the existing data from the original entity. As we're working on supporting custom fields on both Orders and Products, doing this migration will make the naming more correct.


As a new model **115** was already created from a previous PR https://github.com/woocommerce/woocommerce-ios/pull/13724/ , this migration uses that same model and does not create a new one.

The migration follows the lightweight migration outlined [in Apple's documentation](https://developer.apple.com/documentation/coredata/migrating_your_data_model_automatically), which uses the  renaming ID method. A more visual guide for that is available, which I followed: https://stackoverflow.com/a/68693108

The task turned out to require plenty of changes that I found difficult to split to smaller PRs without making build error, so instead I'll try to explain each commits:

1. 82c2e48a449f918c51385386ee985c52ffec0106 : This commit simply does the lightweight migration in the Xcode UI following the guide above.
2. 4faf9620e38c9026a4fde4d676a3e424f654bc14 : [In a previous PR](https://github.com/woocommerce/woocommerce-ios/pull/13620/files#r1717838003), I discussed with @itsmeichigo about keeping `public typealias OrderMetaData = Networking.MetaData` instead of renaming it to `MetaData`, hoping to not have to touch anything in Yosemite. The plan was to then create a new typealias for `ProductMetaData`. As it turned out, the `ReadOnlyConvertible` for`Storage.MetaData` is connected to the Yosemite model, and I don't see an easy way to have one `ReadOnlyConvertible` supporting two different type aliases, so this commit renames the Yosemite model to `MetaData`.
3. 3ebc5adf01228a62c20002251acd7ec806f8d3d6 : This is the bulk of the PR. After doing lightweight migration, turns out there are multiple manual changes to be done, such as:
   - Generating the `CoreDataClass` and `CoreDataProperties` for `MetaData` and removing/renaming the old one.
   - Updating `ReadOnlyConvertible` as mentioned in step 2 above.
   - Updating various places in Storage that still refers to `OrderMetaData`.

The rest of the commits I hope are pretty self-explanatory and does not require further discussions.

## Steps to reproduce

1. Don't build yet, but instead have an Order with custom fields added to it,
2. Build and run the app on `trunk`, and find that Order, ensure its custom fields are shown in **Order Details -> View Custom Fields**
3. Build and run the app on the current branch, wait for the migration to finish and ensure that it doesn't crash upon the migration. 
4. Check the **Order Details -> View Custom Fields** again and ensure it's shown properly.
5. Finally, the migration test should also pass on CI.

## Testing information

Built and ran the app on simulator iPhone 15 Pro iOS 17.5.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

hafizrahman committed:
Lightweight migration to rename entity `OrderMetaData` to `MetaData`

This follows the guide in:
https://developer.apple.com/documentation/coredata/migrating_your_data_model_automatically
https://stackoverflow.com/a/68693108

hafizrahman committed:
Rename typealias in model for OrderMetaData to MetaData.

hafizrahman committed:
Various required manual updates (classes, function calls, types) to follow the entity renaming.

hafizrahman committed:
Add migration test.

hafizrahman committed:
Add migration note.

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
fix files path

hafizrahman committed:
Update unit tests post entity migration.

itsmeichigo commented on Storage/Storage/Tools/StorageType+Extensions.swift:
```diff
@@ -749,4 +742,13 @@ public extension StorageType {
+    /// Retrieves the Stored Metadata.
+    ///
+    func loadMetaData(siteID: Int64, metadataID: Int64) -> MetaData? {
+        let predicate = \MetaData.order?.siteID == siteID && \MetaData.metadataID == metadataID
```
Just curious: do we need to update this predicate if later we need to get metadata from products?

hafizrahman replied:
Yeah, this PR is only specifically for migrating the Entity while keeping everything the same (for now).

There's a separate PR to add Product attribute to the Entity https://github.com/woocommerce/woocommerce-ios/issues/13751, and then this function will need to be updated there as you said to check product, too.

itsmeichigo commented on Storage/StorageTests/CoreData/MigrationTests.swift:
```diff
@@ -2952,6 +2952,49 @@ final class MigrationTests: XCTestCase {
+        let migratedOrder = try XCTUnwrap(targetContext.first(entityName: "Order"))
+        let migratedMetaData = try XCTUnwrap(migratedOrder.value(forKey: "customFields") as? NSSet)
+        XCTAssertEqual(migratedMetaData.count, 1)
+        XCTAssertTrue(migratedMetaData.anyObject() is NSManagedObject)
```
❓ Is it possible to compare the attribute values with the old one to ensure that old data has been migrated successfully?

hafizrahman replied:
Good idea, added that in https://github.com/woocommerce/woocommerce-ios/pull/13736/commits/8c67bab3f36efe45edcf54aa1af5dc1d9a82f15d

itsmeichigo reviewed (approved):
I tested with an order with custom fields and confirmed that the migration works as expected - the custom fields stay intact after the migration. I'd suggest adding the same instruction on the PR description and leaving an `[Internal]` note on the release notes so that the change can be beta tested.

I also left some non-blocking question/suggestion in the comments below.

hafizrahman committed:
Add old and new data comparison to ensure the attributes are migrated correctly

hafizrahman commented:
I'll wait for @pmusolino to chime in before merging this 🙏🏼

hafizrahman committed:
Merge branch 'trunk' into task/13619-rename-ordermetadata-entity-to-metadata

* trunk: (91 commits)
  Remove Equatable from CardPresentPaymentsModalButtonViewModel
  Remove View+WithoutAnimation
  Update PointOfSaleCardPresentPaymentInLineMessage.swift
  Update PointOfSaleCardPresentPaymentTapSwipeInsertCardMessageView.swift
  Animate TotalsView when paymentState changes
  Synchronize icon, title, and message animations of all in line message views
  13691 Use title and body fonts on connection views
  13691 Remove unused view
  13705 Update modal corner radius to match designs
  13705 Update modal shadow to match designs
  13705 Update POS modal overlay/shadow to designs
  13730 Update string in test to match fixed punctuation
  13730 Add error message fixes to 20.2 release notes
  Improve punctuation of card payment errors
  13386 Update tests to match new stop/start logic
  13386 Prevent reader disconnection after card tap
  Disable animation for card reader background view
  Animate transitions between different status views
  Add an extension to disable animation for specific views
  Make PointOfSaleCardPresentPaymentMessageType equatable
  ...

hafizrahman committed:
Update release notes.

pmusolino reviewed (approved):
Thanks, Hafiz, for implementing this.

As far as I remember, I have never done a migration like this before (changing the model name) using the lightweight migration outlined [in Apple's documentation](https://developer.apple.com/documentation/coredata/migrating_your_data_model_automatically). According to the guidelines, such changes should be done using Mapping Models P91TBi-48i. Looking at the Core Data model, this is the first time this approach has been used in the codebase.

For example, when renaming `Attribute` to `GenericAttribute` from model 31 to model 32, a Mapping Model was created for this purpose. However, for this specific case, the lightweight migration might be sufficient. The downside is that errors can be harder to pinpoint since the migration process is more opaque and automatic, and it’s not super clear where the migration happens. For this reason, I strongly suggest adding a reference in the `MIGRATIONS` file, pointing to the lightweight migration.

Another point to consider is to add this model and also the missing model No. 114 inside the test `test_it_can_load_the_expected_model_versions` under `ManagedObjectModelsInventoryTests`.

pmusolino committed:
Merge commit 'be0b83249041c1898dcb8692a58fafa06299be36' into task/13619-rename-ordermetadata-entity-to-metadata

pmusolino committed:
Merge branch 'trunk' into task/13619-rename-ordermetadata-entity-to-metadata

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/221">20.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Add inferring check in the unit test for the migration

hafizrahman committed:
Merge branch 'trunk' into task/13619-rename-ordermetadata-entity-to-metadata

hafizrahman committed:
Add more information about the entity renaming.

hafizrahman committed:
Update with the latest two models.

hafizrahman commented:
@pmusolino  Thank you for sharing your thoughts.

> However, for this specific case, the lightweight migration might be sufficient.

I understand this concern. From reading the documentation, it seems like this is exactly the kind of thing that lightweight migration supports. To be extra safe, I added a check in https://github.com/woocommerce/woocommerce-ios/pull/13736/commits/662796b0d7f15cb3e0105e6344058aef9abd05fc , inspired by the [Confirming whether Core Data can infer the model](https://developer.apple.com/documentation/coredata/migrating_your_data_model_automatically#2903987) section in the documentation.

>  For this reason, I strongly suggest adding a reference in the MIGRATIONS file, pointing to the lightweight migration.

This is done in https://github.com/woocommerce/woocommerce-ios/pull/13736/commits/91107f57e4e7139185cd8a09f1c0416754dbe4cb

> Another point to consider is to add this model and also the missing model No. 114 inside the test test_it_can_load_the_expected_model_versions under ManagedObjectModelsInventoryTests.

Good point, added in https://github.com/woocommerce/woocommerce-ios/pull/13736/commits/a5535e818c4d1b0e99911f594a8e38cd4b8d29b2

Please have a final look and let me know if anything's missing. Given that this is already approved, I'll merge this early tomorrow morning (my time) before code freeze unless something new comes up.

pmusolino commented:
Looks good to me! Thanks, @hafizrahman.
